### PR TITLE
Convert codebase to asynchronous code (reroll v2)

### PR DESCRIPTION
**Pull request by drzraf**
Reroll and continuation  of #354

### CODE REVIEW

**Comment by AidasK** [comment]
Looks great. Code is much more simpler and more readable

**Comment by ilessiivi** [comment]
Thanks for the effort!

Two quick notes:
1. The `.include()` in [Eventizer.js](https://github.com/flowjs/flow.js/pull/363/files#diff-4ca72adb73f499c201c33074dda2efb10e0305e1b806ca0fd98bf006e0337205R320) needs to be `.includes()`
2. A couple of the tests fail, I'll take a look if it's just some missing `await`s or something more

I'll close #354, let's continue here.

File: test/setupSpec.js [hunk 1]
```diff
@@ -34,26 +32,26 @@ describe('setup', function() {
     expect(flow.opts.simultaneousUploads).toBe(flow.defaults.simultaneousUploads);
   });
 
-  it('should show methods initial state', function() {
-    expect(flow.uploadNextChunk()).toBe(false);
+  it('should show methods initial state', async function() {
+    expectAsync(flow.uploadNextChunk()).toBe(false);
```
Comment (ilessiivi) [inline test/setupSpec.js:36]: `toBeResolvedTo(false)` is needed with expectAsync instead of regular `toBe`

File: test/fileAddSpec.js [hunk 2]
```diff
@@ -151,12 +151,12 @@ describe('fileAdd event', function() {
...
       });
 
-      var flowfiles = await flow.asyncAddFiles([
+      var flowfiles = await flow.addFiles([
         new File(['aaa'], 'aaa.bin'),
         new File(['GGG'], 'GGG.bin'),
       ]);
       expect(flowfiles[0].name).toBe('bbb.bin');
-      expect(await flowfiles[0].file.text()).toBe('xxx');
+      expectAsync(flowfiles[0].file.text()).toBe('xxx');
```
Comment (ilessiivi) [inline test/fileAddSpec.js:159]: `toBeResolvedTo('xxx')` is needed with expectAsync instead of regular `toBe`

File: src/Flow.js [hunk 3]
```diff
@@ -633,33 +544,31 @@ export default class Flow extends Eventizer {
...
    */
-  async asyncAddFiles(fileList, event = null, initFileFn = this.opts.initFileFn) {
+  async addFiles(fileList, event = null, initFileFn = this.opts.initFileFn) {
     let item, file, flowfile, uniqueIdentifier, states = [];
     const iterator = this.filterFileList(fileList, event);
 
     while ((item = iterator.next()) && !item.done) {
       [file, uniqueIdentifier] = item.value;
-      if (! await this.aHook('filter-file', file, event)) {
+      if (! await this.hook('filter-file', file, event)) {
```
Comment (ilessiivi) [inline src/Flow.js:555]: When a file is added, the `filter-file` hook is run twice: once without await in `*filterFileList` generator and then with await in the iterator. I think `async addFiles` should look more like [this](https://github.com/flowjs/flow.js/blob/ad86c184b3da3cccaeb9438460166ce444457105/src/Flow.js#L519). (I'd also really like to have the simpler for loop in `addFiles` and remove the generator entirely.)

File: src/Eventizer.js [hunk 4]
```diff
@@ -358,7 +317,6 @@ export default class extends EventTarget {
     const returns = await Promise.all(calls.map(e => e.apply(this, args)));
 
     this.emitCatchAll(name, ...args);
```
Comment (ilessiivi) [inline src/Eventizer.js:319]: The `catch-all` event won't be emitted when there is no custom callback due to the `return` above it. Probably the best way to correct this is to add a second `this.emitCatchAll(name, ...args);` inside the `if (! calls.length) {` block. Although that results in repeated code, it preserves the old order in which the hook and event are fired: the hook's custom callback first and catch-all last.

We also discussed earlier about removing the catch-all event when there is a custom callback. I'm fine with that too, but it is a breaking change.

The `emitCatchAll` call could also be moved before the `if` to keep the code clean, but it could introduce an edge case bug for a user that depended on the event order, so this is the least desirable fix in my opinion. (I can add a test spec for this, unless we end up removing the catch-all firing from hooks with custom callbacks.)

Comment (ilessiivi) [reply src/Eventizer.js:319]: Let me propose one more rewrite of `async hook`:
```javascript
async hook(name, ...args) {
  const calls = this._hooks[name] || [],
        isFilter = this.isFilter(name);

  let returns = isFilter ? true : args[0];

  if (calls.length) {
    // console.log(`[event] Fire ${calls.length} async hook for "${name}"${args.length ? ' with ' + args.length + ' arguments' : ''}`);
    const results = await Promise.all(calls.map(e => e.apply(this, args)));

    returns = isFilter ? !results.includes(false) : results;
  }

  this.emitCatchAll(name, ...args);

  return returns;
}
```
With the `if` this way, the promise handling stays for bound hooks only and catch-all fires always last.

File: test/fileAddSpec.js [hunk 5]
```diff
@@ -12,59 +12,59 @@ describe('fileAdd event', function() {
...
+  it('should validate file-added', async function() {
     spyOn(console, 'warn');
     flow.on('file-added', (file) => {
       delete file.file;
       return false;
     });
-    flow.addFile(new Blob(['file part']));
+    await flow.addFile(new Blob(['file part']));
     expect(flow.files.length).toBe(0);
     expect(console.warn).toHaveBeenCalledTimes(1);
```
Comment (ilessiivi) [inline test/fileAddSpec.js:69]: This should be the last failing test. I think the expected warning is for filtering in `file-added` hook.

Comment (ilessiivi) [inline test/fileAddSpec.js:69]: This should be the last failing test. I think the expected warning is for filtering in `file-added` hook.

Comment (drzraf) [reply test/fileAddSpec.js]: Removing the console.warn altogether. Backward-compatibility is, indeed, non-sense by now :)

File: src/Eventizer.js (hunk 4 above)
Comment (drzraf) [reply src/Eventizer.js:319]: Taken.

File: src/Flow.js (hunk 3 above)
Comment (drzraf) [reply src/Flow.js:550]: We could indeed simplify the loop now but also pass `readFileFn` as an argument, etc...
But let's postpone this for now to just merge something passing the testsuite with the minimum possible diff/change to ease review.

**Comment by drzraf** [comment]
I'll try to remove `asyncReadFileFn` in this very PR, but either way I think it's now mergeable (so we can work on other PR related to `generateUniqueIdentifier`, or other enhancements you had in mind @ilessiivi ...)
Do you confirm?

**Comment by ilessiivi** [comment]
It accomplishes the original goals and it passes all the tests - I would be happy to have this PR merged! Indeed there are still a couple of little things we should do in smaller individual PRs, but I think the project is soon ready for a v3 pre-release.

**Comment by AidasK** [comment]
image
@drzraf Do you know why I am getting these emails? What is being deployed with this approve?

**Comment by AidasK** [comment]
Sadly this email does not have a link to a branch or MR which requires this review
